peer/brontide.go

```diff
@@ -358,6 +365,57 @@ type Config struct {
 	Quit chan struct{}
 }
 
+// The pingManager is a structure that is designed to manage the internal state
+// of the ping pong lifecycle with the remote peer. We assume there is only one
+// ping outstanding at once.
+type pingManager struct {
+	// pingTime is a rough estimate of the RTT (round-trip-time) between us
+	// and the connected peer. This time is expressed in microseconds.
+	// To be used atomically.
+	// TODO(roasbeef): also use a WMA or EMA?
+	pingTime atomic.Int64
+
+	// pingLastSend is the Unix time expressed in nanoseconds when we sent
+	// our last ping message. To be used atomically.
+	pingLastSend atomic.Int64
+
+	// We keep track of the current size of the requested pong payload.
+	// This value can only validly range from [0,65531]. Any value < 0
+	// is interpreted as if there is no outstanding ping message.
+	outstandingPongSize atomic.Int32
+}
+
+// markPingSend is responsible for updating all of the bookkeeping inside of
+// the pingManager after the ping message has been written to the wire.
+func (m *pingManager) markPingSend(pongSize uint16) {
+	now := time.Now().UnixNano()
+	m.pingLastSend.Store(now)
+	m.outstandingPongSize.Store(int32(pongSize))
+}
+
+// cancelPing is used by the timeout logic. If there was an outstanding ping to
+// cancel then this function will return true. If there is no outstanding ping
+// then we return false.
+func (m *pingManager) cancelPing() bool {
+	v := m.outstandingPongSize.Swap(-1)
+	return v >= 0
+}
+
+// resolvePing should be called whenever we receive a pong message from the
+// remote peer, if the payload we receive from them is congruent with the
+// outstandingPongSize, then we return true. A false return value indicates
+// a failure to respond to the ping correctly.
+func (m *pingManager) resolvePing(payload []byte) bool {
+	if len(payload) != int(m.outstandingPongSize.Swap(-1)) {
+		return false
+	}
+	pingSendTime := m.pingLastSend.Load()
+	delay := (time.Now().UnixNano() - pingSendTime) / 1000
```

Comment:
if we leave this as a `time.Duration`, then we can just use this method: https://pkg.go.dev/time#Duration.Microseconds